### PR TITLE
fix: [PL-32111]: Query optimization for list projects call

### PR DESCRIPTION
**Pull request by nishantsaini20**
## Describe your changes
Project list call depends on fetching the list of permitted projects. Earlier the criteria was formed in the following way:
```
{
  "accountIdentifier": "OzzP3vZfT2qXhoCa_fy84w",
  "$or": [
    {
      "orgIdentifier": "cypress_org_Dinesh_DeBruijn",
      "identifier": "cypress_project_Magdalena_Meyer"
    },
    {
      "orgIdentifier": "cypress_org_Dinesh_DeBruijn",
      "identifier": "main_cypress_project_Hong_Jimenez"
    },
    {
      "orgIdentifier": "cypress_org_Unnur_Dijkstra",
      "identifier": "cypress_project_Steinunn_Mitchell"
    }
  ]
}
```
With increase in number of projects, the `$or` clause after having more than 100 entries ignores index and hence the query was not performant resulting into timeouts as well. This solved by querying on _id field as below:
```
{
  "$match": {
    "_id": {
      "$in": [
        "ObjectId("641847b742fd232b50e3aeb1"),
        "ObjectId("6418558c42fd232b50e477a4"),
        "ObjectId("641959e422533020bc2c6976"),
        "ObjectId("6418a3a15744f53d153cd00b"),
        "ObjectId("6418482e42fd232b50e3b2d3")
      ]
    }
  }
}
```
The above query uses `_id_` index and is far better performant.

**Testing:**
1. Create projects in shared org.
2. Created shared projects in non shared org
3. Created projects not shared with anyone.
4. Logged in with user to which projects are share.
5. Logged in user should be able to view only those projects which are shared with the user.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47070)

### CODE REVIEW

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47070)* status: 2 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/test/java/io/harness/ng/core/impl/ProjectServiceImplTest.java` line 402 at r1](https://reviewable.io/reviews/harness/harness-core/47070#-NUR88LkDveyQmtklRy2:-NUR88LkDveyQmtklRy3:b-ofee5v) ([raw file](https://github.com/harness/harness-core/blob/711d483bc71ae74a9b6739786c7951ad7a03b598/120-ng-manager/src/test/java/io/harness/ng/core/impl/ProjectServiceImplTest.java#L402)):*
> ```Java
>   @Owner(developers = NISHANT)
>   @Category(UnitTests.class)
>   public void testlistPermittedProjects() {
> ```

@nishantsaini20 Can we avoid mocking here and test using in-memory mongo.

Basically create projects in mongo, query and then assert result



_Code quote (from 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java):_
```Java
Scope
```

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47070)* status: 2 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 477 at r1](https://reviewable.io/reviews/harness/harness-core/47070#-NUR8ObhBsr67lJwR575:-NUR8Obi6v_tqiUb6_dQ:btxbf61) ([raw file](https://github.com/harness/harness-core/blob/711d483bc71ae74a9b6739786c7951ad7a03b598/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L477)):*
> ```Java
>     List<Project> projects = projectRepository.findAll(criteria);
> 
>     Map<Scope, String> scopeToProjectIdMap = projects.stream().collect(Collectors.toMap(project
> ```

Do Scope implements hashCode and equals correctly. Asking because its the key in the map

**Comment by nishantsaini20** [comment]
trigger messagecheck

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47070)* status: 3 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 477 at r1](https://reviewable.io/reviews/harness/harness-core/47070#-NUR8ObhBsr67lJwR575:-NUV0_k15F1chAkNDE6p:b-4xjh9u) ([raw file](https://github.com/harness/harness-core/blob/711d483bc71ae74a9b6739786c7951ad7a03b598/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L477)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Do Scope implements hashCode and equals correctly. Asking because its the key in the map
</blockquote></details>

No. It was't. Added EqualsAndHashCode annotation to Scope class.
___
*[`120-ng-manager/src/test/java/io/harness/ng/core/impl/ProjectServiceImplTest.java` line 402 at r1](https://reviewable.io/reviews/harness/harness-core/47070#-NUR88LkDveyQmtklRy2:-NUV0mwU6-qHg5zojh_S:bupx9cq) ([raw file](https://github.com/harness/harness-core/blob/711d483bc71ae74a9b6739786c7951ad7a03b598/120-ng-manager/src/test/java/io/harness/ng/core/impl/ProjectServiceImplTest.java#L402)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

@nishantsaini20 Can we avoid mocking here and test using in-memory mongo.

Basically create projects in mongo, query and then assert result
</blockquote></details>

As discussed, let's take it as a separate task.

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47070#-:-NUVAPaR5Jub_7hqfUGP:bnfp4nl)* status: 3 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47070#-:-NUVARP2-10RjArH3qfB:bnfp4nl)* status: 3 file(s) to review. 1 of 2 approvals obtained.
